proposals/0330-collection-conditionals.md

```diff
@@ -36,13 +40,168 @@ The most notable use case for this is conditional inclusion of tests for the Swi
 
 ## Proposed solution
 
-The solution proposed is to allow #if conditionals using their exiting syntax inside collection literals surrounding sublists of elements. These elements would be either included or not included in the resulting array or dictionary instance dependant on the truth of the `#if`, `#elseif` or `#else` i.e. whether they where "active". One new syntactic requirement is the trailing comma in sublists before or inside conditional clauses is not optional as it would normally be at the end of the collection.
+The solution proposed is to allow `#if` conditionals using their exiting syntax inside collection literals surrounding sublists of elements. These elements would be either included or not included in the resulting array or dictionary instance dependant on the truth of the `#if`, `#elseif` or `#else` i.e. whether they where "active".
 
 ## Detailed design
 
-The implementation involves a slight modification to `Parser::parseList` to detect `#if` "statements" if they are present and call `Parser::parseIfConfig` recursively call `parseList` to gather the elements in the clauses of the conditionals. Only the elements in the "active" clause of the conditional are included in the elements of the final `CollectionExpr` AST instance after parsing.
+### Grammar changes
+
+This proposal changes `array-literal-items` and `dictionary-literal-items` to accept `conditional-compilation-block` in it.
+
+```diff
+- array-literal-items → array-literal-item ','(opt) | array-literal-item ',' array-literal-items
++ array-literal-items → array-literal-item ','(opt) | array-literal-item ',' array-literal-items | conditional-compilation-block array-literal-items
+
+- dictionary-literal-items → dictionary-literal-item ','(opt) | dictionary-literal-item ',' dictionary-literal-items
++ dictionary-literal-items → dictionary-literal-item ','(opt) | dictionary-literal-item ',' dictionary-literal-items | conditional-compilation-block dictionary-literal-items
```

Comment:
There don't *have* to be any items after the conditional compilation block, right? So we probably want this to be:

```suggestion
- array-literal-items → array-literal-item ','(opt) | array-literal-item ',' array-literal-items
+ array-literal-items → array-literal-item ','(opt) | array-literal-item ',' array-literal-items | conditional-compilation-block array-literal-items(opt)

- dictionary-literal-items → dictionary-literal-item ','(opt) | dictionary-literal-item ',' dictionary-literal-items
+ dictionary-literal-items → dictionary-literal-item ','(opt) | dictionary-literal-item ',' dictionary-literal-items | conditional-compilation-block dictionary-literal-items(opt)
```